stab_controller_multi_lqr_node.cpp

```diff
@@ -0,0 +1,416 @@
+    ReadInMatrixArray("/home/ascento/catkin_ws/src/ascento/Matlab/DynMod/fixed_drive/ros_compatible/k_matrix_");
+    // Read in Imu offsets
+    ReadInDoubleArray("/home/ascento/catkin_ws/src/ascento/Matlab/DynMod/fixed_drive/ros_compatible/imu_offset");
+  }
+  void StabControllerMultiLqrNode::InitSubPub() {
+    //Initialise subscribers
+    imu_sub = nh.subscribe("/est_states/imu", 1, &StabControllerMultiLqrNode::SaveImuCallback, this);
+    anydrive_left_sub = nh.subscribe("/sens_states/hip_mot/left", 1, &StabControllerMultiLqrNode::SaveAnydriveLeftCallback, this);
+    anydrive_right_sub = nh.subscribe("/sens_states/hip_mot/right", 1, &StabControllerMultiLqrNode::SaveAnydriveRightCallback, this);
+
+    maxon_execution_pub = nh.advertise<ascento_msgs::wheel_input_torque>("/actuate/wheel_mot", 1);
+  }
+  void StabControllerMultiLqrNode::InitSrv() {
+    //Initialise services
+    controller_starter = nh.advertiseService("/stab_control_start", &StabControllerMultiLqrNode::StartStopCallback, this);
+    imu_autotune = nh.advertiseService("/set_imu_offset", &StabControllerMultiLqrNode::SetImuOffsetCallback, this);
+  }
+  bool StabControllerMultiLqrNode::ReadParameters() {
+    // Geometric parameters
+    if (!nh.getParam("/wheel/radius", wheel_radius)) return false;
+    if (!nh.getParam("/phi/min", phi_min)) return false;
+    if (!nh.getParam("/phi/max", phi_max)) return false;
+    if (!nh.getParam("/maxon_ec90_max_torque", saturation_ec90)) return false;
+    if (!nh.getParam("/encoder/quadcounts", encoder_quadcounts)) return false;
+    if (!nh.getParam("/measured_imu_offset", measured_imu_offset)) ROS_INFO("No param \"/measured_imu_offset\" found, run srv \"/set_imu_offset\"");
+
+    // Controller Parameters
+    if (!nh.getParam("/stab_controller_multi_lqr/sampling_rate", sampling_rate)) return false;
+
+    return true;
+  }
+
+  // Threading methods maxon motors
+  void StabControllerMultiLqrNode::ReadLeft(){
+    ros::Rate rate_ReadLeft(sampling_rate);
+    while (ros::ok()) {
+      if(is_running){
+        std::cout << "/* readleft */" << '\n';
+        meas_left(2) = (maxon_left.GetPosition()*2.0*M_PI/encoder_quadcounts-encoder_offset_left);
+        meas_left(3) = maxon_left.GetAverageVelocity()*2.0*M_PI/60.0;
+        rate_ReadLeft.sleep();
+      }
+    }
+  }
+  void StabControllerMultiLqrNode::ReadRight(){
+    ros::Rate rate_ReadRight(sampling_rate);
+    while (ros::ok()) {
+      if(is_running){
+        std::cout << "/* readright */" << '\n';
+        meas_right(2) = (-maxon_right.GetPosition()*2.0*M_PI/encoder_quadcounts-encoder_offset_right);
+        meas_right(3) = -maxon_right.GetAverageVelocity()*2.0*M_PI/60.0;
+        rate_ReadRight.sleep();
+      }
+    }
+  }
+  void StabControllerMultiLqrNode::WriteLeft(){
+    ros::Rate rate_WriteLeft(sampling_rate);
+    while (ros::ok()) {
+      if(is_running){
+        std::cout << "/* writeleft */" << '\n';
+        maxon_left.MoveWithTorque(input_left(0));
```

Comment:
I feel that you also need to have a look at the docs for the motor api to determine how thread safe they are. The biggest issue is what happens if you try to read and write to the same object at the same time. There may be some internal shared state that is not thread safe.

I now feel this may be the most likely cause of the issues you are having, as both commands will likely cause the system to send some information to the motor, if the calls are not thread-safe the two messages could be spliced together during transmission resulting in undefined behaviour.